AsyncDisplayKit/Details/ASDataController.mm

```diff
@@ -391,30 +317,26 @@ - (void)_reloadDataWithAnimationOptions:(ASDataControllerAnimationOptions)animat
       
       void (^transactionBlock)() = ^{
         LOG(@"Edit Transaction - reloadData");
-        
-        // Remove everything that existed before the reload, now that we're ready to insert replacements
-        NSArray *indexPaths = ASIndexPathsForMultidimensionalArray(_editingNodes[ASDataControllerRowNodeKind]);
-        [self _deleteNodesAtIndexPaths:indexPaths withAnimationOptions:animationOptions];
-        
-        NSMutableArray *editingNodes = _editingNodes[ASDataControllerRowNodeKind];
-        NSMutableIndexSet *indexSet = [[NSMutableIndexSet alloc] initWithIndexesInRange:NSMakeRange(0, editingNodes.count)];
-        [self _deleteSectionsAtIndexSet:indexSet withAnimationOptions:animationOptions];
-        
-        [self willReloadData];
-        
-        // Insert each section
-        NSMutableArray *sections = [NSMutableArray arrayWithCapacity:sectionCount];
-        for (int i = 0; i < sectionCount; i++) {
-          [sections addObject:[[NSMutableArray alloc] init]];
-        }
-        
-        [self _insertSections:sections atIndexSet:[NSIndexSet indexSetWithIndexesInRange:NSMakeRange(0, sectionCount)] withAnimationOptions:animationOptions];
-        
-        [self _batchLayoutNodes:updatedNodes atIndexPaths:updatedIndexPaths withAnimationOptions:animationOptions];
-        
-        if (completion) {
-          dispatch_async(dispatch_get_main_queue(), completion);
+
+        NSMutableArray *editingNodes = [NSMutableArray arrayWithCapacity:sectionCount];
+
+        for (NSUInteger i = 0; i < sectionCount; i++) {
+          [editingNodes addObject:[NSMutableArray array]];
         }
+
+        ASInsertElementsIntoMultidimensionalArrayAtIndexPaths(editingNodes, updatedIndexPaths, updatedNodes);
+        _editingNodes[ASDataControllerRowNodeKind] = editingNodes;
+
+        [self batchLayoutNodes:updatedNodes ofKind:ASDataControllerRowNodeKind atIndexPaths:updatedIndexPaths];
+
+        [self commitChangesToNodesOfKind:ASDataControllerRowNodeKind withCompletion:^{
```

Comment:
Small optimization here: can pass `NIL` here if `_delegateDidReloadData` is `NO` and `completion` is `nil`. That way we don't schedule a block on main thread that essentially does nothing.